src/mastic.rs

```diff
@@ -0,0 +1,331 @@
+use crate::{
+    codec::{CodecError, Decode, Encode, ParameterizedDecode},
+    field::FieldElement,
+    flp::{
+        szk::{Szk, SzkProofShare},
+        Type,
+    },
+    vdaf::{
+        poplar1::Poplar1AggregationParam,
+        xof::{Seed, Xof},
+        Aggregatable, Client, Vdaf, VdafError,
+    },
+    vidpf::{
+        Vidpf, VidpfInput, VidpfKey, VidpfPublicShare, VidpfServerId, VidpfValue, VidpfWeight,
+    },
+};
+use std::{fmt::Debug, io::Cursor, marker::PhantomData};
+/// The main struct implementing the Mastic VDAF.
+/// Composed of a shared zero knowledge proof system and a verifiable incremental
+/// distributed point function.
+#[derive(Clone, Debug)]
+pub struct Mastic<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    algorithm_id: u32,
+    szk: Szk<T, P, SEED_SIZE>,
+    pub(crate) vpf: Vidpf<VidpfWeight<T::Field>, 16>,
+    /// The maximum length of the private label associated with any input.
+    pub bits: usize,
+    phantom: PhantomData<P>,
+}
+
+impl<T, P, const SEED_SIZE: usize> Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    ///Creates a new instance of Mastic, with a specific label length and validity type.
+    pub fn new(
+        algorithm_id: u32,
+        szk: Szk<T, P, SEED_SIZE>,
+        vpf: Vidpf<VidpfWeight<T::Field>, 16>,
+        bits: usize,
+    ) -> Self {
+        Self {
+            algorithm_id,
+            szk,
+            vpf,
+            bits,
+            phantom: PhantomData,
+        }
+    }
+}
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation, a set of prefixes to evaluate at that level,
+/// and, optionally, the aggregate results of prior levels.
+pub type MasticAggregationParam = Poplar1AggregationParam;
+
+/// Contains broadcast information shared between parties to support VIDPF correctness.
+pub type MasticPublicShare<V> = VidpfPublicShare<V>;
+
+/// Add necessary traits for MasticPublicShare here
+
+/// Message sent by the [`Client`] to each Aggregator during the Sharding phase.
+#[derive(Clone, Debug)]
+pub struct MasticInputShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// VIDPF key share.
+    vidpf_key: VidpfKey,
+
+    /// The proof share.
+    proofs_share: SzkProofShare<F, SEED_SIZE>,
+}
+impl<'a, T, P, const SEED_SIZE: usize> ParameterizedDecode<(&'a Mastic<T, P, SEED_SIZE>, usize)>
+    for MasticInputShare<T::Field, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    fn decode_with_param(
+        decoding_parameter: &(&'a Mastic<T, P, SEED_SIZE>, usize),
```

Comment:
You can unpack these arguments here, making the code more readable
```suggestion
        (mastic, whatever_this_thing_is_called): &(&'a Mastic<T, P, SEED_SIZE>, usize),
```